src/test/java/org/jitsi/videobridge/cc/vp8/VP8AdaptiveTrackProjectionTest.java

```diff
@@ -0,0 +1,858 @@
+                initialState, logger);
+
+        PacketInfo firstPacketInfo = generator.nextPacket();
+        Vp8Packet firstPacket = firstPacketInfo.packetAs();
+
+        for (int i = 0; i < 11; i++)
+        {
+            PacketInfo packetInfo = generator.nextPacket();
+            Vp8Packet packet = packetInfo.packetAs();
+
+            assertFalse(context.accept(packetInfo, packet.getTemporalLayerIndex(), 2));
+        }
+
+        assertFalse(context.accept(firstPacketInfo, firstPacket.getTemporalLayerIndex(), 2));
+
+        for (int i = 0; i < 30; i++)
+        {
+            PacketInfo packetInfo = generator.nextPacket();
+            Vp8Packet packet = packetInfo.packetAs();
+
+            assertFalse(context.accept(packetInfo, packet.getTemporalLayerIndex(), 2));
+        }
+
+        generator.requestKeyframe();
+
+        for (int i = 0; i < 9958; i++)
+        {
+            PacketInfo packetInfo = generator.nextPacket();
+            Vp8Packet packet = packetInfo.packetAs();
+
+            assertTrue(context.accept(packetInfo, packet.getTemporalLayerIndex(), 2));
+            context.rewriteRtp(packetInfo);
+        }
+    }
+
+    @Test
+    public void twoStreamsNoSwitchingTest() throws RewriteException
+    {
+        Vp8PacketGenerator generator1 = new Vp8PacketGenerator(3);
+        Vp8PacketGenerator generator2 = new Vp8PacketGenerator(3);
+        generator2.setSsrc(0xdeadbeefL);
+
+        DiagnosticContext diagnosticContext = new DiagnosticContext();
+        diagnosticContext.put("test", "twoStreamsNoSwitchingTest");
+
+        RtpState initialState =
+            new RtpState(1, 10000, 1000000);
+
+        VP8AdaptiveTrackProjectionContext context =
+            new VP8AdaptiveTrackProjectionContext(diagnosticContext, payloadType,
+                initialState, logger);
+
+        int expectedSeq = 10001;
+        long expectedTs = 1003000;
+        for (int i = 0; i < 10000; i++)
+        {
+            PacketInfo packetInfo1 = generator1.nextPacket();
+            Vp8Packet packet1 = packetInfo1.packetAs();
+
+            assertTrue(context.accept(packetInfo1, packet1.getTemporalLayerIndex() + 3, 5));
+
+            PacketInfo packetInfo2 = generator2.nextPacket();
+            Vp8Packet packet2 = packetInfo2.packetAs();
+
+            assertFalse(context.accept(packetInfo2, packet2.getTemporalLayerIndex(), 5));
+
+            context.rewriteRtp(packetInfo1);
+
+            assertEquals(expectedSeq, packet1.getSequenceNumber());
+            assertEquals(expectedTs, packet1.getTimestamp());
+            expectedSeq = RtpUtils.applySequenceNumberDelta(expectedSeq, 1);
+
+            if (packet1.isEndOfFrame())
+            {
+                expectedTs = RtpUtils.applyTimestampDelta(expectedTs, 3000);
+            }
+        }
+    }
+
+    @Test
+    public void twoStreamsSwitchingTest() throws RewriteException
+    {
+        Vp8PacketGenerator generator1 = new Vp8PacketGenerator(3);
```

Comment:
i'm wondering if it would be useful (and also if it's in scope for this PR) to have a test that feeds the code with real VP8 simulcast (stored in a pcap file, for instance)?

I seem to recall that the engine produces (and the pacer transmits) frames together, so you have frame A (packets x1,x2,x3) followed by frame B (packets y1,y2,y3) and so on and so forth, which is not captured by these (admittedly very much needed) test cases.

This could be useful for debugging purposes as well, i.e. capture ingress simulcast that we broke in the egress and play it back to see what went wrong; and it would enable us to easily write a test case to avoid the bug in the future.

Reply:
I agree it seems like a useful tool.  Do you think it needs to be part of this PR, though?

Reply:
No, it doesn't have to be, especially if we feel confident enough that the refactor doesn't introduce any significant regressions.